### PR TITLE
Change data properties in `@fluid-tools/benchmark`

### PR DESCRIPTION
**Pull request by jikim-msft**
#### Description

[7825](https://dev.azure.com/fluidframework/internal/_queries/edit/7825/?triage=true)

While working on 7825, found that the `@fluid-tools/benchmark` custom data object is missing a property which will ultimately log `undefined` in Kusto. 

Also capitalized the property name to conform to the memory benchmark test.

### CODE REVIEW

File: tools/benchmark/src/runBenchmark.ts [hunk 1]
```diff
@@ -191,16 +191,18 @@ class BenchmarkState<T> implements BenchmarkTimer<T> {
...
+				"Iterations per Batch": this.iterationsPerBatch,
 			},
 			customDataFormatters: {
-				"batch count": (v: unknown): string => prettyNumber(v as number, 0),
-				"period (ns/op)": (v: unknown) => prettyNumber(1e9 * (v as number), 2),
-				"relative margin of error": (v: unknown) => `±${(v as number).toFixed(2)}%`,
-				"iterations per batch": (v: unknown) => `${prettyNumber(v as number, 0)}`,
+				"Batch Count": (v: unknown): string => prettyNumber(v as number, 0),
+				"Period (ns/op)": (v: unknown) => prettyNumber(1e9 * (v as number), 2),
+				"Margin of Error": (v: unknown): string => prettyNumber(v as number, 0),
```
Comment (alexvy86) [inline tools/benchmark/src/runBenchmark.ts:203]: We want 2 decimals for this, I think this will print it with no decimals?

Comment (jikim-msft) [reply tools/benchmark/src/runBenchmark.ts:203]: Thanks for a good catch 😄 .

**Review by alexvy86 — APPROVED** [review]
Note that the changes in casing will have to be reflected in the other PR, where we consume these changes and look for these keys.

**Comment by jikim-msft** [comment]
Memory Test: 

@jikim-msft ➜ .../FluidFramework/experimental/dds/attributable-map (benchmark/fix-reporter) $ pnpm run test:memory-profiling:report

> @fluid-experimental/attributable-map@2.2.0 test:memory-profiling:report /workspaces/FluidFramework/experimental/dds/attributable-map
> mocha --config ./src/test/memory/.mocharc.cjs


SharedMap memory usage
Results file: /workspaces/FluidFramework/experimental/dds/attributable-map/.memoryTestsOutput/SharedMap_memory_usage_perfresult.json
status  name                                          Heap Used Avg  Margin of Error  Relative Margin of Error  Heap Used StdDev  Iterations
------  --------------------------------------------  -------------  ---------------  ------------------------  ----------------  ----------
    ✔   Create empty map                                   1,680.34           ±14.47                     ±0.86            160.88         500
    ✔   Add 1000 integers to a local map                 143,364.41        ±3,583.14                     ±2.50         46,787.31         689
    ✔   Add 1000 integers to a local map, clear it        -2,071.67          ±910.08                    ±43.93         19,357.50        1829
    ✔   Add 10000 integers to a local map              1,872,748.94       ±10,742.06                     ±0.57         37,573.41          50
    ✔   Add 10000 integers to a local map, clear it        2,095.46           ±52.23                     ±2.49            448.27         298
    ✔   Add 100000 integers to a local map            16,941,356.26          ±325.88                     ±0.00          1,139.86          50
    ✔   Add 100000 integers to a local map, clear it       2,666.38           ±27.68                     ±1.04             96.81          50



Overall summary

status  suite name              # of passed tests  total time (s)
------  ----------------------  -----------------  --------------
    ✔   SharedMap memory usage         7 out of 7            67.1
        total                          7 out of 7            67.1

**Comment by jikim-msft** [comment]
Time Execution Test:

@jikim-msft ➜ .../FluidFramework/packages/dds/tree (benchmark/fix-reporter) $ pnpm run test:benchmark:report

> @fluidframework/tree@2.2.0 test:benchmark:report /workspaces/FluidFramework/packages/dds/tree
> mocha --exit --perfMode --parentProcess --fgrep @Benchmark --reporter @fluid-tools/benchmark/dist/MochaReporter.js --timeout 60000

Writing test results relative to package to nyc/junit-report.xml and nyc/junit-report.json

Direct Object
Results file: /workspaces/FluidFramework/node_modules/.pnpm/file+tools+benchmark/node_modules/@fluid-tools/benchmark/dist/.output/Direct_Object_perfresult.json
status  name                                Batch Count  Period (ns/op)            Margin of Error          Relative Margin of Error  Iterations per Batch
------  ----------------------------------  -----------  ------------------------  -----------------------  ------------------------  --------------------
    ✔   clone JS Object: 'canada'                     7      0.003434457616071429   0.00003391353153879582        0.9874494121021785                    16
    ✔   sum JS Object: 'canada'                       5      0.007810926849999999   0.00007109318026683098        0.9101759833640098                     8
    ✔   averageValues JS Object: 'canada'             7     0.0009184979776785715  0.000007684529270954835        0.8366408481787685                    64
    ✔   clone JS Object: 'twitter'                   11      0.040027518318181826   0.00038608137421339793        0.9645398726556249                     2
    ✔   sum JS Object: 'twitter'                      5           0.0023074247875  0.000009238535010450646       0.40038293167770894                    32
    ✔   averageValues JS Object: 'twitter'           14  0.0000056220160696847095    5.3779782733814936e-8         0.956592476208112                 16384

```
{
            "benchmarkName": "sum JS Object: 'canada'",
            "elapsedSeconds": 0.378745029,
            "customData": {
                "Batch Count": 5,
                "Period (ns/op)": 0.007810926849999999,
                "Margin of Error": 0.00007109318026683098,
                "Relative Margin of Error": 0.9101759833640098,
                "Iterations per Batch": 8
            },
            "customDataFormatters": {}
        },
```

**Comment by jikim-msft** [comment]
https://github.com/microsoft/FluidFramework/blob/0a4e642f3b776de5f9f7cf85538cfb6041307786/tools/benchmark/src/mocha/customOutputRunner.ts#L41

`customDataFormatters` is not storing any value here, and I believe this could be a reason why. Will try to include the change in this PR.

**Comment by alexvy86** [comment]
Yes, we should keep the current formatting for that output, and ideally also the column ordering, so it should look like this (of note, currently we have a "total time (s)" column that is missing in the new output for execution time tests, let's bring that back):

```
Direct Object
Results file: /home/alex/code/FluidFramework/node_modules/.pnpm/@fluid-tools+benchmark@0.48.0/node_modules/@fluid-tools/benchmark/dist/.output/Direct_Object_perfresult.json
status  name                                period (ns/op)  relative margin of error  iterations per batch  batch count  total time (s)
------  ----------------------------------  --------------  ------------------------  --------------------  -----------  --------------
    ✔   clone JS Object: 'canada'             3,502,552.47                    ±1.13%                    16           88            5.01
    ✔   sum JS Object: 'canada'               7,038,451.23                    ±1.46%                     8           88            5.02
    ✔   averageValues JS Object: 'canada'       890,289.07                    ±6.29%                   128           43            5.01
    ✔   clone JS Object: 'twitter'           20,243,425.75                    ±4.39%                     2          122            5.03
    ✔   sum JS Object: 'twitter'              2,022,632.23                    ±2.15%                    32           77            5.05
    ✔   averageValues JS Object: 'twitter'        4,801.44                    ±4.46%                16,384           63            5.04
```

**Comment by jikim-msft** [comment]
`pnpm run test:memory`

`SharedMap_memory_usage_perfresult.json`

```json
{
    "suiteName": "SharedMap memory usage",
    "benchmarks": [
        {
            "benchmarkName": "Create empty map",
            "elapsedSeconds": 6.40703022,
            "customData": {
                "Heap Used Avg": 1698.0042105263158,
                "Heap Used StdDev": 189.00447604582394,
                "Margin of Error": 16.997355426134494,
                "Relative Margin of Error": 1.0010196276760686,
                "Iterations": 500
            },
            "customDataFormatters": {
                "Heap Used Avg": "1,698.00",
                "Heap Used StdDev": "189.00",
                "Margin of Error": "±17.00",
                "Relative Margin of Error": "±1.00",
                "Iterations": "500"
            }
        },    
    ]
}
```

`pnpm run test:benchmark:report`

`Direct_Object_perfresult.json`

```json
{
    "suiteName": "Direct Object",
    "benchmarks": [
        {
            "benchmarkName": "clone JS Object: 'canada'",
            "elapsedSeconds": 1.2569377,
            "customData": {
                "Batch Count": 20,
                "Iterations per Batch": 16,
                "Period (ns/op)": 0.0037011110124999997,
                "Margin of Error": 0.00003682178879856034,
                "Relative Margin of Error": 0.9948847433702949
            },
            "customDataFormatters": {
                "Batch Count": "20",
                "Iterations per Batch": "16",
                "Period (ns/op)": "3,701,111.01",
                "Margin of Error": "±0.00%",
                "Relative Margin of Error": "±0.99%"
            }
        },
    ]
}
```

**Comment by jikim-msft** [comment]
In addition, 

I found scripts `test:memory` & `test:memory-profiling:report` have identical command `mocha --config src/test/memory/.mocharc.cjs`. Should I still leave this?

**Review by seanimam — APPROVED** [review]

File: tools/benchmark/src/ResultTypes.ts [hunk 2]
```diff
@@ -64,7 +64,7 @@ export interface BenchmarkData {
 	/**
 	 * Custom data formatters.
```
Comment (Josmithr) [inline tools/benchmark/src/ResultTypes.ts:65]: We should document the expected format of this. What are the keys? What are the values?

File: tools/benchmark/src/ResultTypes.ts [hunk 3]
```diff
@@ -64,7 +64,7 @@ export interface BenchmarkData {
...
 	 */
-	customDataFormatters: Record<string, (value: unknown) => string>;
+	customDataFormatters: Record<string, string>;
```
Comment (Josmithr) [inline tools/benchmark/src/ResultTypes.ts:67]: Is this intended to be reassigned after being set? If not
```suggestion
	readonly customDataFormatters: Record<string, string>;
```

**Review by Josmithr — APPROVED** [review]
In order to publish a new version of this for use by the rest of the repo, you'll need to bump the version number and update `CHANGELOG.md`.

I left a couple of minor comments, but otherwise, the API changes look good to me 🙂

**Comment by jikim-msft** [comment]
`packages/dds/tree/Direct_Object_perfresult.json`

```json
{
    "suiteName": "Direct Object",
    "benchmarks": [
        {
            "benchmarkName": "clone JS Object: 'canada'",
            "Batch Count": 80,
            "Iterations per Batch": 16,
            "Period (ns/op)": 3856568.296874999,
            "Margin of Error": 0.00006049487918598355,
            "Relative Margin of Error": 1.5686194183311346
        },
        {
            "benchmarkName": "sum JS Object: 'canada'",
            "Batch Count": 9,
            "Iterations per Batch": 8,
            "Period (ns/op)": 7863358.999999998,
            "Margin of Error": 0.00007225853335498965,
            "Relative Margin of Error": 0.918927055918338
        },
        {
            "benchmarkName": "averageValues JS Object: 'canada'",
            "Batch Count": 5,
            "Iterations per Batch": 64,
            "Period (ns/op)": 939232.1812499999,
            "Margin of Error": 0.000004862195561145792,
            "Relative Margin of Error": 0.5176777008082093
        },
        {
            "benchmarkName": "clone JS Object: 'twitter'",
            "Batch Count": 95,
            "Iterations per Batch": 2,
            "Period (ns/op)": 26123052.31578948,
            "Margin of Error": 0.001661469120269007,
            "Relative Margin of Error": 6.360164578718737
        },
        {
            "benchmarkName": "sum JS Object: 'twitter'",
            "Batch Count": 5,
            "Iterations per Batch": 32,
            "Period (ns/op)": 2387740.8874999997,
            "Margin of Error": 0.000019856424448717328,
            "Relative Margin of Error": 0.831598795022826
        },
        {
            "benchmarkName": "averageValues JS Object: 'twitter'",
            "Batch Count": 53,
            "Iterations per Batch": 16384,
            "Period (ns/op)": 5729.89070215765,
            "Margin of Error": 8.767663342477144e-8,
            "Relative Margin of Error": 1.5301624059208652
        }
    ]
}
```

Only contains `rawData` and no longer contains `formattedValue` in the output file. 

Screenshot 2024-07-31 at 17 56 29

Renders prettified version in the console output as expected.

**Comment by jikim-msft** [comment]
Run `pnpm run test:memory` in `packages/dds/map`: 

```json
{
    "suiteName": "SharedDirectory memory usage",
    "benchmarks": [
        {
            "benchmarkName": "Create empty directory",
            "Heap Used Avg": 1710.938947368421,
            "Heap Used StdDev": 213.37509429141096,
            "Margin of Error": 19.189028707853225,
            "Relative Margin of Error": {
                "arithmeticMean": 1710.938947368421,
                "variance": 45528.93086386852,
                "standardDeviation": 213.37509429141096,
                "marginOfError": 19.189028707853225,
                "standardErrorOfMean": 9.79032076931287,
                "samples": [
                    1640,
                    1640,
                    1640,
                    1640,
                    DELTED FOR SPACE
                ],
                "marginOfErrorPercent": 1.1215495875739863
            },
            "Iterations": 500
        },
```
Screenshot 2024-07-31 at 18 02 59

Same format rendered as above.

File: tools/benchmark/src/ResultTypes.ts (hunk 3 above)
Comment (jikim-msft) [reply tools/benchmark/src/ResultTypes.ts]: `customDataFormatters` removed so no longer applies 😄 .

File: tools/benchmark/src/ResultTypes.ts (hunk 2 above)
Comment (jikim-msft) [reply tools/benchmark/src/ResultTypes.ts]: `customDataFormatters` removed so no longer applies 😄 .

**Comment by jikim-msft** [comment]
@Josmithr 

I've updated the changelog for `0.50.0`. But I believe the bump to `0.51.0` in `package.json` should be done in a separate PR, am I correct?

1. Merge this PR containing the necessary changes for `0.50.0`
2. Create a separate PR using `flub` command which bumps the version in `benchmark/package.json` to `0.51.0` [example](https://github.com/microsoft/FluidFramework/pull/22024).
3. Create a `release/benchmark/0.50.0` branch in the upstream and run the release of `0.50.0`.

**Pull request merged**